### PR TITLE
[Create] Error handling

### PR DESCRIPTION
**Pull request by phpadventure**

### CODE REVIEW

File: src/infrastructure/Application.php [hunk 1]
```diff
@@ -72,20 +73,46 @@ public function handle(Request $request,
...
+    private function handleException(Request $request,int $type, \Exception $exception): Response
+    {
+        $event = new GetResponseForExceptionEvent($this, $request, $type, $exception);
+        $this->dispatcher->dispatch(KernelEvents::EXCEPTION, $event);
+
+        // a listener might have replaced the exception
+        $exception = $event->getException();
+
+        if (! $event->hasResponse()) {
+            throw $exception;
```
Comment (Antonyan) [inline src/infrastructure/Application.php]: Dangerous:
It could be uncaught Exception

File: src/infrastructure/Application.php [hunk 2]
```diff
@@ -72,20 +73,46 @@ public function handle(Request $request,
...
             $this->eventDispatcher->dispatch('request', new RequestEvent($request, $controller[0], $controller[1]));
 
-            $response = \call_user_func_array($controller, $arguments);
-        } catch (ResourceNotFoundException $exception) {
-            $response = new Response('Not Found', 404);
+            $response = call_user_func_array($controller, $arguments);
         } catch (Exception $exception) {
-            $response = new Response('An error occurred', 500);
+
+            if ($catch === false) {
```
Comment (Antonyan) [inline src/infrastructure/Application.php]: Complicated:
Is there a simpler way?

File: src/infrastructure/Application.php (hunk 1 above)
Comment (phpadventure) [reply src/infrastructure/Application.php]: you are right, but than this means that we didn't attached any listener to the exception event. Could be change to 500 http error.

Comment (phpadventure) [reply src/infrastructure/Application.php]: It's infrastracture error. not application.

File: src/infrastructure/Application.php (hunk 2 above)
Comment (phpadventure) [reply src/infrastructure/Application.php]: Yes, not to use default symfony exception listener, for which we decided to give a try. This is their default implementation how they handle subrequests.

File: src/infrastructure/Application.php (hunk 1 above)
Comment (Antonyan) [reply src/infrastructure/Application.php]: Still dangerous:
we need to add at least two types of behavior for prod and dev env

File: src/infrastructure/Application.php [hunk 3]
```diff
@@ -72,20 +73,46 @@ public function handle(Request $request,
...
             $controller = $this->controllerResolver->getController($request);
             $arguments = $this->argumentResolver->getArguments($request, $controller);
 
             $this->eventDispatcher->dispatch('request', new RequestEvent($request, $controller[0], $controller[1]));
 
-            $response = \call_user_func_array($controller, $arguments);
-        } catch (ResourceNotFoundException $exception) {
-            $response = new Response('Not Found', 404);
+            $response = call_user_func_array($controller, $arguments);
         } catch (Exception $exception) {
```
Comment (Antonyan) [inline src/infrastructure/Application.php]: Maintainability: 
We need to decorate Symfony resource not found exception

File: src/app/Services/Error.php [hunk 4]
```diff
@@ -2,21 +2,23 @@
...
-     * @param FlattenException $exception
-     * @return Response
-     * @throws InvalidArgumentException
-     */
-    public function exception(ResourceNotFoundException $exception)
+    public function handle(FlattenException $exception)
     {
-        $msg = 'Something went wrong! ('.$exception->getMessage().')';
-        return new Response($msg, $exception->getStatusCode());
+        if ($exception->getClass() == ResourceNotFoundException::class) {
```
Comment (Antonyan) [inline src/app/Services/Error.php]: Code complexity: 
Rid off exception name check

**Pull request merged**